crates/red_knot_python_semantic/src/types.rs

```diff
@@ -4658,6 +4665,346 @@ impl<'db> GeneralCallableType<'db> {
+                            // checked against the variadic parameter in `self`. This loop does
+                            // that by only moving the `other` iterator forward.
+                            loop {
+                                let Some(other_parameter) = parameters.peek_other() else {
+                                    break;
+                                };
+                                if !matches!(
+                                    other_parameter.kind(),
+                                    ParameterKind::PositionalOnly { .. }
+                                        | ParameterKind::Variadic { .. }
+                                ) {
+                                    // Any other parameter kind cannot be checked against a
+                                    // variadic parameter and is deferred to the next iteration.
+                                    break;
+                                }
+                                if !is_subtype(
+                                    other_parameter.annotated_type(),
+                                    self_parameter.annotated_type(),
+                                ) {
+                                    return false;
+                                }
+                                parameters.next_other();
+                            }
+                        }
+
+                        (ParameterKind::Variadic { .. }, ParameterKind::Variadic { .. }) => {
+                            if !is_subtype(
+                                other_parameter.annotated_type(),
+                                self_parameter.annotated_type(),
+                            ) {
+                                return false;
+                            }
+                        }
+
+                        (
+                            _,
+                            ParameterKind::KeywordOnly { .. }
+                            | ParameterKind::KeywordVariadic { .. },
+                        ) => {
+                            // Keyword parameters are not considered in this loop as the order of
+                            // parameters is not important for them and so they are checked by
+                            // doing name-based lookups.
+                            break;
+                        }
+
+                        _ => return false,
+                    }
+                }
+            }
+        }
+
+        // At this point, the remaining parameters in `other` are keyword-only or keyword variadic.
+        // But, `self` could contain any unmatched positional parameters.
+        let (self_parameters, other_parameters) = parameters.into_remaining();
+
+        // Collect all the keyword-only parameters and the unmatched standard parameters.
+        let mut self_keywords = HashMap::new();
+
+        // Type of the variadic keyword parameter in `self`.
+        let mut self_keyword_variadic = None;
```

Comment:
I think I need to change this up because then there's no way to differentiate between keyword-variadic parameter being absent and that the annotated type is absent.

Reply:
Done in https://github.com/astral-sh/ruff/pull/16804/commits/12ff6b9243317fed7ac7a947d3bf6951caf5b84b